### PR TITLE
FHAC-519 Incorporate AsynchronousEJB for DocumentRetrieve Services

### PR DESCRIPTION
**Pull request by vdmehta06**
for RD, I am able to audit one entry for Initiating and one entry for responding gateway for both Standard and Passthrough mode.
RD codebase is refactored and old Orchestration pattern is removed

### CODE REVIEW

File: Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/BaseInboundDocRetrieve.java [hunk 1]
```diff
@@ -57,16 +59,28 @@
...
         InboundDocRetrieveOrchestratable orchResponse = (InboundDocRetrieveOrchestratable) getOrchestrator().process(
-                inboundOrchestrable);
+            inboundOrchestrable);
+        auditResponse(body, orchResponse.getResponse(), assertion, webContextProperties);
+
         return orchResponse.getResponse();
     }
+
+    protected DocRetrieveAuditLogger getLogger() {
+        return new DocRetrieveAuditLogger();
```
Comment (christophermay07) [inline Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/BaseInboundDocRetrieve.java]: Do we need a new instance each time?

File: Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieve.java [hunk 2]
```diff
@@ -43,32 +42,30 @@
 public class PassthroughInboundDocRetrieve extends BaseInboundDocRetrieve {
 
     private final PolicyTransformer pt;
-    private final AuditTransformer at;
     private final InboundDelegate ad;
     private final CONNECTInboundOrchestrator orch;
+    DocRetrieveAuditLogger docRetrieveAuditLogger = new DocRetrieveAuditLogger();
```
Comment (christophermay07) [inline Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieve.java]: Two questions:
1. Is default access okay for this, or should it be private?
2. Do we even need to hold this?  Class inherits "getLogger()" from super.

File: Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/nhin/InboundDocRetrieveStrategyImpl.java [hunk 3]
```diff
@@ -121,9 +121,9 @@ public RetrieveDocumentSetResponseType sendToAdapter(InboundDocRetrieveOrchestra
     public void auditInboundResponseMessage(InboundDocRetrieveOrchestratable message) {
 
         LOG.debug("Calling audit log for doc retrieve response received from adapter (a0)");
-        docRetrieveLogger.auditResponseMessage(message.getRequest(), message.getResponse(), message.getAssertion(),
-            null, NhincConstants.AUDIT_LOG_INBOUND_DIRECTION, NhincConstants.AUDIT_LOG_ADAPTER_INTERFACE,
-            Boolean.FALSE, message.getWebContextProperties(), NhincConstants.DOC_RETRIEVE_SERVICE_NAME);
+//        docRetrieveLogger.auditResponseMessage(message.getRequest(), message.getResponse(), message.getAssertion(),
```
Comment (christophermay07) [inline Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/nhin/InboundDocRetrieveStrategyImpl.java]: Always follow this rule when submitting a PR with commented-out code:

Make sure to add a comment above the code as to why it wasn't removed.  If you can't come up with a reason, remove the code instead.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/CONNECTOrchestrationBase.java [hunk 4]
```diff
@@ -168,28 +163,6 @@ protected Orchestratable createErrorResponse(OutboundOrchestratable message, Str
...
-
-        if (message != null && message.getAuditTransformer() != null) {
-            AuditTransformer transformer = message.getAuditTransformer();
-            transformer.transformResponse(message);
-            
-        }
-        return resp;
-    }
-
     private AcknowledgementType audit(LogEventRequestType message, AssertionType assertion) {
```
Comment (alameluchidambaram) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/CONNECTOrchestrationBase.java]: This method seems unused. Can you check if it's used?

File: Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java [hunk 5]
```diff
@@ -51,6 +56,22 @@
...
+    RetrieveDocumentSetRequestType retrieveDocumentSetRequestType;
+    RetrieveDocumentSetResponseType retrieveDocumentSetResponseType;
+
+    @Before
+    public void setup() {
+        retrieveDocumentSetRequestType = mock(RetrieveDocumentSetRequestType.class);
+        retrieveDocumentSetResponseType = mock(RetrieveDocumentSetResponseType.class);
+        webContextProp = mock(Properties.class);
+        nhinTargetSystemType = mock(NhinTargetSystemType.class);
+        assertionType = mock(AssertionType.class);
```
Comment (alameluchidambaram) [inline Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java]: Are these mock objects used?

File: Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/nhin/InboundDocRetrieveStrategyImpl.java (hunk 3 above)
Comment (vdmehta06) [inline Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/nhin/InboundDocRetrieveStrategyImpl.java]: ok. Removed unused and commented code

File: Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java (hunk 5 above)
Comment (vdmehta06) [inline Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java]: Removed unwanted mock objects from Inbound and Outbound side as well.

File: Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieve.java (hunk 2 above)
Comment (vdmehta06) [inline Product/Production/Services/DocumentRetrieveCore/src/main/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieve.java]: I removed this - instance creation statement as this is no longer used as - getLogger() method gets a instance of audit logger.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/AbstractOrchestrable.java [hunk 6]
```diff
@@ -44,9 +44,8 @@ protected AbstractOrchestrable() {
 
     }
 
-    protected AbstractOrchestrable(PolicyTransformer pt, AuditTransformer at) {
+    protected AbstractOrchestrable(PolicyTransformer pt) {
         policyTransformer = pt;
-        auditTransformer = at;
     }
 
     public AuditTransformer getAuditTransformer() {
```
Comment (christophermay07) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/AbstractOrchestrable.java]: Do we still use the AuditTransformer for anything?  I can't find where the getter or setter is used.

**Comment by TabassumJafri** [comment]
PR looks good. Just wanted to know the purpose of so many Unsupported Methods in the code

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/AbstractPassthroughOrchestratable.java [hunk 7]
```diff
@@ -45,12 +44,13 @@ protected AbstractPassthroughOrchestratable() {
      * @param at audit transformer
      * @param ad
      */
```
Comment (alameluchidambaram) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/orchestration/AbstractPassthroughOrchestratable.java:44]: Can you remove these  params in javadoc for auditTransformer and ad ?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryOrchestratable.java [hunk 8]
```diff
@@ -75,27 +72,19 @@ public OutboundDocQueryOrchestratable(OutboundDelegate delegate) {
...
-     * @param a
-     *            assertion.
-     * @param name
-     *            serviceName.
-     * @param t
-     *            target.
-     * @param r
-     *            AdhocQUery Request.
+     * @param d Delegate.
+     * @param at AuditTransformer.
```
Comment (alameluchidambaram) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryOrchestratable.java]: Can you remove this param ?

File: Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryOrchestratable_a0.java [hunk 9]
```diff
@@ -56,27 +55,20 @@ public OutboundDocQueryOrchestratable_a0() {
...
-     * @param a
-     *            assertion.
-     * @param name
-     *            serviceName.
-     * @param t
-     *            target.
-     * @param req
-     *            AdhocQUery Request.
+     * @param d Delegate.
+     * @param at AuditTransformer.
```
Comment (alameluchidambaram) [inline Product/Production/Services/DocumentQueryCore/src/main/java/gov/hhs/fha/nhinc/docquery/entity/OutboundDocQueryOrchestratable_a0.java]: same as above comment.

**Comment by alameluchidambaram** [comment]
Reviewed PR and approve for merge after clean up of Javadoc params for AuditTransformer

File: Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java [hunk 10]
```diff
@@ -59,7 +75,7 @@ public void invoke() {
         Properties webContextProperties = new Properties();
 
         InboundDocRetrievePolicyTransformer_g0 pt = new InboundDocRetrievePolicyTransformer_g0();
-        InboundDocRetrieveAuditTransformer_g0 at = new InboundDocRetrieveAuditTransformer_g0();
+//        InboundDocRetrieveAuditTransformer_g0 at = new InboundDocRetrieveAuditTransformer_g0();
```
Comment (alameluchidambaram) [inline Product/Production/Services/DocumentRetrieveCore/src/test/java/gov/hhs/fha/nhinc/docretrieve/inbound/PassthroughInboundDocRetrieveTest.java]: Do you need this commented line in future?

**Comment by alameluchidambaram** [comment]
Reviewed PR and good with merge.

**Comment by alameluchidambaram** [comment]
Tabassum- The AuditTransformer was used only for Document Retrieve service and there was an orchestrator used across all services where the AuditTransformer was orchestrated. Since it was not used by other service this was implemented as an unsupported method.

**Pull request merged**